File: 프로그래머스/python/LV3/네트워크.py
from collections import deque


def bfs(i, n, computers, visited):
    q = deque([i])
    visited[i] = 1
    while q:
        k = q.popleft()
        for j in range(n):
            if computers[k][j] == 1 and visited[j] == 0:
                visited[j] = 1
                q.append(j)
    return 1


def solution(n, computers):
    answer = 0
    visited = [0 for i in range(n)]

    for i in range(n):
        if visited[i] == 0:
            answer += bfs(i, n, computers, visited)
    return answer

File: 프로그래머스/javascript/LV3/네트워크.js
function solution(n, computers) {
  let answer = 0;
  const visited = Array(n).fill(false);

  for (let i = 0; i < n; i++) {
    // 첫번째 컴퓨터부터 연결된 모든 컴퓨터 탐색
    if (visited[i]) continue;

    const queue = [i];
    visited[i] = true;
    let head = 0;

    while (queue.length > head) {
      const x = queue[head++];

      for (let y = 0; y < n; y++) {
        if (visited[y] || !computers[x][y]) continue;
        queue.push(y);
        visited[y] = true;
      }
    }

    answer++;
  }

  return answer;
}
